proposals/0000-uncurated-layer.rst

```diff
@@ -0,0 +1,57 @@
+
+The aim of this proposal is to separate these two purposes, by allowing authors to distinguish if they wish to opt-out of following the PVP and the attendant curation process that helps to maintain correct dependency information. In so doing, they recognize that their discoverability on Hackage may be reduced, since the UI will give precedence to those packages which, in conjunction with potential cabal revisions, are self-contained with regards to the information necessary to build them.
+
+Proposed Change
+---------------
+
+This is a phased proposal with gradual rollout possible, in dependency order of the steps. The first phase is immediate. Other phases may be rolled out at any pace, but necessarily in the order specified (including the collapse of two steps into a joint step).
+
+The desired end-state will have the following properties:
+
+1) Packages will have an additional flag set in the Hackage package database, that indicates if they are curated or not. This flag is set *per version*.
+2) Package authors will set this flag *on upload*, by setting the "x-curation" property of the cabal file of a package. If no "x-uncurated" property is set, this will be considered "curated". Along with "curated" and "uncurated," other sub-settings will be made available. In particular: "uncurated-no-trustee-contact" and "uncurated-seeking-adoption".
+3) Hackage will provide two package repository roots -- http://hackage.haskell.org and http://uncurated.hackage.haskell.org These roots will provide index-01.tar.gz files that contain the information, respectively, for curated packages, or for all packages. The uncurated root will contain no revision information for uncurated packages (even those that have been adopted), but will contain revision information for curated packages.
+4) Curated packages cannot depend on uncurated packages, and the hackage server will detect this as an error at upload time.
+5) Uncurated packages may be "adopted" into the curated ecosystem by trustees. Metadata revisions necessarily remove the x-uncurated property from the revised cabal metadata.
+
+The first phase of this rollout is simply social. It has the following properties:
+
++ Hackage trustees will recognize and respect the uncurated flag. When it is set to "uncurated-no-trustee-contact" they will not contact authors with any issues. They *will* retain the ability to make metadata revisions, bearing in mind that they must remove the x-uncurated property from revised metadata.
```

Comment:
This concerns me. I maintain several packages. Most of them would end up with `x-curation: uncurated-no-trustee-contact`. A Hackage trustee could revise some version of one of my packages, effectively curating it. That means a consumer of the curated package index would get a package tarball that I didn't author. For example, someone might try to `cabal install lackey-1.0.0`, which doesn't support `servant-0.14`, but end up with a revised package that does support `servant-0.14` (since no code changes were required). 

This situation seems similar to when a non-maintainer adds a package to Stackage, but I think it's different enough to be concerning. The worst case scenario for Stackage is that a user ends up with an old version of a package. On Hackage, with this proposal, a user could end up with a package that none of the package authors or maintainers ever even knew about.